#### Task
Provide a script to collect Google Trends weekly, daily and hourly data with the keyword ‘bitcoin’ since 2015-01-01 and save results in csv or json file

##### Approach :
1. Use pytrends api to get the data.
2. Extract the data from January,2015 to a dataframe.
3. Create new attributes like year, month, day,hour, week from date.
4. Group data based on year, month, day and get the sum of counts to get daily data.
5. Group data based on year, month, day, week and get the sum of counts to get weekly data.

In [62]:
# importing libraries
from pytrends.request import TrendReq
import pandas as pd

In [2]:
pytrends = TrendReq(hl='en-US', tz=360) 
kw_list = ["bitcoin"]

In [5]:
#making the api call per each year and append each year data to the dataframe
df = pd.DataFrame(columns=['bitcoin','isPartial'])
for year in range(2015,2023):
    temp=pytrends.get_historical_interest(kw_list, year_start=year, month_start=1, day_start=1, hour_start=0, year_end=year, month_end=12, day_end=31, hour_end=0, cat=0, sleep=0)
    temp.reset_index();
    df=pd.concat([df,temp])

The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The requ

The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The requ

The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The request failed: Google returned a response with code 429.
The requ

##### few requests are failed with response code 429. response code 429 denotes too many requests in a given time.

In [6]:
df.shape

(15042, 2)

In [16]:
df.reset_index(inplace=True)

In [17]:
df

index bitcoin isPartial
0     2015-01-01 00:00:00      27     False
1     2015-01-01 01:00:00      33     False
2     2015-01-01 02:00:00      77     False
3     2015-01-01 03:00:00      61     False
4     2015-01-01 04:00:00      22     False
...                   ...     ...       ...
15037 2016-09-15 20:00:00      65     False
15038 2016-09-15 21:00:00      92     False
15039 2016-09-15 22:00:00      52     False
15040 2016-09-15 23:00:00      63     False
15041 2016-09-16 00:00:00      68     False

[15042 rows x 3 columns]

In [18]:
#converting index to datetime object and storing it as date column
df['date'] = pd.to_datetime(df['index'])

In [20]:
#dropping unneccessary fields
df.drop(columns=["index","isPartial"], inplace=True)

##### Extracting the year, month, day, hour, week columns

In [23]:
df['year']=pd.DatetimeIndex(df['date']).year

In [24]:
df['month']=pd.DatetimeIndex(df['date']).month

In [25]:
df['day']=pd.DatetimeIndex(df['date']).day

In [29]:
df['hour']=pd.DatetimeIndex(df['date']).hour

In [57]:
df['week']= df['date'].dt.strftime('%U')

In [58]:
df.head(25)

bitcoin                date  year  month  day  hour week  weekday  day_name
0       27 2015-01-01 00:00:00  2015      1    1     0   00        3  Thursday
1       33 2015-01-01 01:00:00  2015      1    1     1   00        3  Thursday
2       77 2015-01-01 02:00:00  2015      1    1     2   00        3  Thursday
3       61 2015-01-01 03:00:00  2015      1    1     3   00        3  Thursday
4       22 2015-01-01 04:00:00  2015      1    1     4   00        3  Thursday
5       25 2015-01-01 05:00:00  2015      1    1     5   00        3  Thursday
6       36 2015-01-01 06:00:00  2015      1    1     6   00        3  Thursday
7       28 2015-01-01 07:00:00  2015      1    1     7   00        3  Thursday
8       36 2015-01-01 08:00:00  2015      1    1     8   00        3  Thursday
9       38 2015-01-01 09:00:00  2015      1    1     9   00        3  Thursday
10      32 2015-01-01 10:00:00  2015      1    1    10   00        3  Thursday
11      37 2015-01-01 11:00:00  2015      1    1    11   00        3  Thursday
12      31 2015-01-01 12:00:00  2015      1    1    12   00        3  Thursday
13      61 2015-01-01 13:00:00  2015      1    1    13   00        3  Thursday
14      38 2015-01-01 14:00:00  2015      1    1    14   00        3  Thursday
15      35 2015-01-01 15:00:00  2015      1    1    15   00        3  Thursday
16      36 2015-01-01 16:00:00  2015      1    1    16   00        3  Thursday
17      58 2015-01-01 17:00:00  2015      1    1    17   00        3  Thursday
18      53 2015-01-01 18:00:00  2015      1    1    18   00        3  Thursday
19      45 2015-01-01 19:00:00  2015      1    1    19   00        3  Thursday
20      38 2015-01-01 20:00:00  2015      1    1    20   00        3  Thursday
21      58 2015-01-01 21:00:00  2015      1    1    21   00        3  Thursday
22      43 2015-01-01 22:00:00  2015      1    1    22   00        3  Thursday
23      66 2015-01-01 23:00:00  2015      1    1    23   00        3  Thursday
24      56 2015-01-02 00:00:00  2015      1    2     0   00        4    Friday

In [64]:
df[['year','month','day','hour','bitcoin']].to_csv("hour_data.csv")

In [39]:
#grouping the data to get daily counts
daily_counts=df.groupby(['year','month','day']).bitcoin.agg([sum])

In [40]:
daily_counts

sum
year month day      
2015 1     1    1014
           2    1107
           3    1230
           4    1426
           5    1409
...              ...
2016 9     12   1535
           13   1406
           14   1314
           15   1490
           16     68

[625 rows x 1 columns]

In [49]:
daily_counts.to_csv('daily_data.csv')

In [61]:
#grouping the data to get weekly counts
weekly_counts=df.groupby(['year','month','week']).bitcoin.agg([sum])
weekly_counts.to_csv('weekly_data.csv')